rust/lancedb/src/remote/client.rs

```diff
@@ -14,42 +14,202 @@
 
 use std::{future::Future, time::Duration};
 
+use log::debug;
 use reqwest::{
     header::{HeaderMap, HeaderValue},
-    RequestBuilder, Response,
+    Request, RequestBuilder, Response,
 };
 
 use crate::error::{Error, Result};
 
+const REQUEST_ID_HEADER: &str = "x-request-id";
+
+#[derive(Debug)]
+pub struct ClientConfig {
+    timeout_config: TimeoutConfig,
+    retry_config: RetryConfig,
+    user_agent: String,
+    // TODO: how to configure request ids?
+}
+
+impl Default for ClientConfig {
+    fn default() -> Self {
+        Self {
+            timeout_config: TimeoutConfig::default(),
+            retry_config: RetryConfig::default(),
+            user_agent: concat!("LanceDB-Rust-Client/{}", env!("CARGO_PKG_VERSION")).into(),
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct TimeoutConfig {
+    /// The timeout for creating a connection to the server.
+    ///
+    /// You can also set the `LANCE_CLIENT_CONNECT_TIMEOUT` environment variable
+    /// to set this value. Use an integer value in seconds.
+    ///
+    /// The default is 120 seconds (2 minutes).
+    pub connect_timeout: Option<Duration>,
+    /// The timeout for reading a response from the server.
+    ///
+    /// You can also set the `LANCE_CLIENT_READ_TIMEOUT` environment variable
+    /// to set this value. Use an integer value in seconds.
+    ///
+    /// The default is 300 seconds (5 minutes).
+    pub read_timeout: Option<Duration>,
+    /// The timeout for keeping idle connections alive.
+    ///
+    /// You can also set the `LANCE_CLIENT_CONNECTION_TIMEOUT` environment variable
+    /// to set this value. Use an integer value in seconds.
+    ///
+    /// The default is 300 seconds (5 minutes).
+    pub pool_idle_timeout: Option<Duration>,
```

Comment:
This is inspired by https://github.com/lancedb/lancedb/blob/c1d9d6f70b757a38dc4bf5671ea95368b86184cb/python/python/lancedb/remote/connection_timeout.py#L32-L33